compilers/concrete-optimizer/concrete-optimizer/src/optimization/dag/multi_parameters/analyze.rs

```diff
@@ -1,38 +1,412 @@
+            compo.apply();
+        }
+    }
+
+    /// Propagates the noise downward in the graph.
+    fn forward_noise(&mut self) -> bool {
+        // We save the old variance to compute the diff at the end.
+        let old_variances = self.variances.clone();
+        let nb_partitions = self.partitions.nb_partitions;
+
+        // We loop through the operators and propagate the noise.
+        for mut operator in self.get_operators_iter_mut() {
+            // Inputs are already computed
+            if operator.operator().is_input() {
+                continue;
+            }
+            let max_var = |acc: SymbolicVariance, input: SymbolicVariance| acc.max(&input);
+            // Operator variance will be used to override the noise
+            let mut operator_variance = OperatorVariance::nan(nb_partitions);
+            // We first compute the noise in the partition of the operator
+            operator_variance[operator.partition().instruction_partition] = match operator
+                .operator()
+                .operator
+            {
+                Operator::Input { .. } => unreachable!(),
+                Operator::Lut { .. } => SymbolicVariance::after_pbs(
+                    nb_partitions,
+                    operator.partition().instruction_partition,
+                ),
+                Operator::LevelledOp { manp, .. } => {
+                    let max_var = operator
+                        .get_inputs_iter()
+                        .map(|a| a.variance()[operator.partition().instruction_partition].clone())
+                        .reduce(max_var)
+                        .unwrap();
+                    max_var.after_levelled_op(*manp)
+                }
+                Operator::Dot {
+                    kind: DotKind::CompatibleTensor { .. },
+                    ..
+                } => todo!("TODO"),
+                Operator::Dot {
+                    kind: DotKind::Unsupported { .. },
+                    ..
+                } => panic!("Unsupported"),
+                Operator::Dot {
+                    inputs,
+                    weights,
+                    kind: DotKind::Simple | DotKind::Tensor | DotKind::Broadcast { .. },
+                } => {
+                    let variance_iterator: Box<dyn Iterator<Item = OperatorVariance>> =
+                        if inputs.len() == 1 {
+                            Box::new(std::iter::repeat(
+                                operator
+                                    .get_inputs_iter()
+                                    .next()
+                                    .unwrap()
+                                    .variance()
+                                    .clone(),
+                            ))
+                        } else {
+                            Box::new(operator.get_inputs_iter().map(|n| n.variance().clone()))
+                        };
+
+                    weights.values.iter().zip(variance_iterator).fold(
```

Comment:
IMO The for loop was clearer. If you want "high level functional ", why not doing it with the Sum trait ?